### PR TITLE
Implemented a SimpleServer

### LINKED ISSUE #116: implement a SimpleServer like puppeteer
by creating a server like https://github.com/GoogleChrome/puppeteer/blob/v1.0.0/test/server/SimpleServer.js we would be able to add routes/auths and more things more easily.

it will remove all of the test-specific routes

this depends on #101

### PR DESCRIPTION
**Pull request by Meir017**
closes #116

### CODE REVIEW

**Comment by Meir017** [comment]
Just a single test failed

https://ci.appveyor.com/project/kblok/puppeteer-sharp/build/1.0.225#L356

only when using `netcoreapp2.0`

🤔

**Comment by kblok** [comment]
@Meir017 the problem I think we are having here is that we run tests in parallels and we have a server in a static field, makes sense?

**Comment by kblok** [comment]
Could your remove the controllers we're not going to use anymore? (like the auth one)

**Comment by kblok** [comment]
I think I found the issue. I have a few ideas for you.

I added a few console.writelines here:
https://github.com/kblok/puppeteer-sharp/pull/119/files#diff-231086e53074993b9c85d09cd4295476R45

and here as:
https://github.com/kblok/puppeteer-sharp/pull/119/files#diff-554899e7b490512df5a6428184dc3e31R18

It turns out that the `Request` object is being changed between one moment and the other
screen shot 2018-04-08 at 2 22 10 pm

I was able to fix it by calling an `action` before the request is changed:

```cs
_requestSubscribers = new ConcurrentDictionary<string, Action<HttpRequest>>();
....
if (_requestSubscribers.TryGetValue(context.Request.Path, out var subscriber))
{
    subscriber(context.Request);
}
```

```cs
public async Task<T> WaitForRequest<T>(string path, Func<HttpRequest, T> selector)
{
    var taskCompletion = new TaskCompletionSource<T>();
    _requestSubscribers.Add(path, (httpRequest) =>
    {
        taskCompletion.SetResult(selector(httpRequest));
    });

    var request = await taskCompletion.Task;
    _requestSubscribers.Remove(path);

    return request;
}
```

**Comment by kblok** [comment]
Nice!

File: lib/PuppeteerSharp.TestServer/SimpleServer.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        public void SetRedirect(string from, string to)
+        {
+            SetRoute(from, context =>
+            {
+                context.Response.Redirect(to);
+                return Task.CompletedTask;
+            });
+        }
+
+        public async Task<T> WaitForRequest<T>(string path, Func<HttpRequest, T> selector)
```
Comment (kblok) [inline lib/PuppeteerSharp.TestServer/SimpleServer.cs]: Is the function a `selector`? That name sounds weird to me.

File: lib/PuppeteerSharp.Tests/Page/SetExtraHttpHeadersTests.cs [hunk 2]
```diff
@@ -15,8 +15,10 @@ public class SetExtraHttpHeadersTests : PuppeteerPageBaseTest
                 ["Foo"] = "Bar"
             });
 
-            var response = await Page.GoToAsync($"{TestConstants.ServerUrl}/headertests/test");
-            Assert.Equal("Bar", await response.TextAsync());
+            var headerTask = Server.WaitForRequest("/empty.html", request => request.Headers["Foo"]);
+            await Task.WhenAll(Page.GoToAsync(TestConstants.EmptyPage), headerTask);
+
+            Assert.Equal("Bar", await headerTask);
```
Comment (kblok) [inline lib/PuppeteerSharp.Tests/Page/SetExtraHttpHeadersTests.cs]: I think `headerTask.Result` it's more clear here. The task was already awaited on 19

File: lib/PuppeteerSharp.TestServer/SimpleServer.cs (hunk 1 above)
Comment (kblok) [reply lib/PuppeteerSharp.TestServer/SimpleServer.cs]: Good point

Comment (kblok) [reply lib/PuppeteerSharp.TestServer/SimpleServer.cs]: Good point

**Comment by kblok** [comment]
This is ready to go right @Meir017?

**Comment by Meir017** [comment]
yup

**Pull request merged**